code/go/0chain.net/smartcontract/dbs/event/snapshot.go

```diff
@@ -44,6 +43,63 @@ type Snapshot struct {
 	BlockCount           int64 `json:"block_count"`                      // Total number of blocks currently
 	AverageTxnFee        int64 `json:"avg_txn_fee"`                      // Average transaction fee per block
 	CreatedAt            int64 `gorm:"autoCreateTime" json:"created_at"` // Snapshot creation date
+	BlobberCount		 int64 `json:"blobber_count"`                    // Total number of blobbers
+	MinerCount			 int64 `json:"miner_count"`                      // Total number of miners
+	SharderCount		 int64 `json:"sharder_count"`                    // Total number of sharders
+	ValidatorCount		 int64 `json:"validator_count"`                  // Total number of validators
+	AuthorizerCount		 int64 `json:"authorizer_count"`                  // Total number of authorizers
+}
+
+func (s *Snapshot) providerCount(provider spenum.Provider) int64 {
+	switch provider {
+	case spenum.Blobber:
+		return s.BlobberCount
+	case spenum.Miner:
+		return s.MinerCount
+	case spenum.Sharder:
+		return s.SharderCount
+	case spenum.Validator:
+		return s.ValidatorCount
+	case spenum.Authorizer:
+		return s.AuthorizerCount
+	default:
+		return 0
+	}
+}
+
+// ApplyDiff applies diff values of global snapshot fields to the current snapshot according to each field's update formula.
+// For some fields, the count of the providers may be needed so a provider parameter is added.
+func (s *Snapshot) ApplyDiff(diff *Snapshot, provider spenum.Provider) {
+	logging.Logger.Debug("SnapshotDiff", zap.Any("provider", provider), zap.Any("diff", diff), zap.Any("snapshot_before", s))
+	s.TotalMint += diff.TotalMint
+	s.TotalChallengePools += diff.TotalChallengePools
+	s.ActiveAllocatedDelta += diff.ActiveAllocatedDelta
+	s.ZCNSupply += diff.ZCNSupply
+	s.TotalValueLocked += diff.TotalValueLocked
+	s.ClientLocks += diff.ClientLocks
+	s.MinedTotal += diff.MinedTotal
+	s.TotalStaked += diff.TotalStaked
+	s.TotalRewards += diff.TotalRewards
+	s.SuccessfulChallenges += diff.SuccessfulChallenges
+	s.TotalChallenges += diff.TotalChallenges
+	s.AllocatedStorage += diff.AllocatedStorage
+	s.MaxCapacityStorage += diff.MaxCapacityStorage
+	s.StakedStorage += diff.StakedStorage
+	s.UsedStorage += diff.UsedStorage
+	s.TransactionsCount += diff.TransactionsCount
+	s.UniqueAddresses += diff.UniqueAddresses
+	s.BlockCount += diff.BlockCount
+
+	if s.TransactionsCount > 0 {
+		s.AverageTxnFee += diff.AverageTxnFee / s.TransactionsCount
+	}
+
+	providerCount := s.providerCount(provider)
+	if providerCount > 0 {
+		s.AverageWritePrice += diff.AverageWritePrice / providerCount
```

Comment:
it can be not correct if it is new blobber, 
previous average was calculated as `sum(wp)/n` and new one as `sum(wp + dif)/ (n+1)` so your share should be `(sum(wp)/n + diff/n)*n/(n+1)` and simply `old_sum * n / (n + 1) + diff / (n + 1)`

Reply:
Nice catch 👌 Fixing it.

Reply:
Fixed. FYI when new providers are added they affect the snapshot count directly as response to the events. https://github.com/0chain/0chain/blob/a89a68aaf0a4cba103f731b35997923d0c0f61a6/code/go/0chain.net/smartcontract/dbs/event/snapshot.go#L318-L354
I added update to the avg write price there to update it whenever a provider is added/removed through the formula `new_avg = total / new_count = old_avg * old_count / new_count`

Reply:
Added some tests and just noticed that I don't need to update average for all the providers, just for the blobbers. Lack of coffee 🙃